File: rust/src/data/httpapi/alerts.rs
use std::time::Duration;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct AlertName(pub String);

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct QueryName(pub String);

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct AlertMessage(pub String);

#[derive(Serialize, Deserialize, Debug, PartialOrd, PartialEq)]
pub struct AlertThresholdValue(pub f64);

#[derive(Serialize, Deserialize, Debug)]
pub struct Alert {
    pub name: AlertName,
    pub query: QueryName,
    #[serde(rename = "intervalSecs", with = "crate::utils::seconds_duration_format")]
    pub interval: Duration,
    #[serde(rename = "repeatIntervalSecs", with = "crate::utils::seconds_duration_format")]
    pub repeat_interval: Duration,
    #[serde(flatten)]
    pub thresholds: AlertThresholds
}

#[derive(Serialize, Deserialize, Debug)]
pub struct AlertThreshold {
    pub message: AlertMessage,
    pub value: AlertThresholdValue,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct AlertThresholds {
    pub warn: AlertThreshold,
    pub critical: AlertThreshold,
}
impl AlertThresholds {
    pub fn state_for(&self, value: &AlertThresholdValue) -> AlertState {
        if value >= &self.critical.value {
            AlertState::Critical { message: &self.critical.message }
        }
        else if value >= &self.warn.value {
            AlertState::Warn { message: &self.warn.message }
        }
        else {
            AlertState::Pass
        }
    }
}

#[derive(Eq, PartialEq, Debug)]
pub enum AlertState<'message> {
    /// [AlertThresholdValue] < [AlertThresholds.warn]
    Pass,
    /// [AlertThresholdValue] >= [AlertThresholds.warn]
    Warn { message: &'message AlertMessage },
    /// [AlertThresholdValue] >= [AlertThresholds.critical]
    Critical { message: &'message AlertMessage },
}

File: go/data/httpapi/alerts.go
package httpapi

import (
	"alert-execution-engine/data"
	. "alert-execution-engine/functional"
	"time"
)

type AlertName string

func (a AlertName) String() string {
	return string(a)
}

type QueryName string

func (q QueryName) String() string {
	return string(q)
}

type AlertMessage string
type AlertThresholdValue float64

type Alert struct {
	Name           AlertName
	Query          QueryName
	Interval       time.Duration
	RepeatInterval time.Duration
	Thresholds     AlertThresholds
}

// JsonAlert is Alert which we parse from JSON.
type JsonAlert struct {
	Name           AlertName         `json:"name"`
	Query          QueryName         `json:"query"`
	Interval       data.DurationSecs `json:"intervalSecs"`
	RepeatInterval data.DurationSecs `json:"repeatIntervalSecs"`
	Warn           AlertThreshold    `json:"warn"`
	Critical       AlertThreshold    `json:"critical"`
}

func (self *JsonAlert) toAlert() Alert {
	return Alert{
		Name:           self.Name,
		Query:          self.Query,
		Interval:       self.Interval.ToDuration(),
		RepeatInterval: self.RepeatInterval.ToDuration(),
		Thresholds: AlertThresholds{
			Warn:     self.Warn,
			Critical: self.Critical,
		},
	}
}

type AlertThreshold struct {
	Message AlertMessage        `json:"message"`
	Value   AlertThresholdValue `json:"value"`
}

//region AlertThresholds

type AlertThresholds struct {
	Warn     AlertThreshold
	Critical AlertThreshold
}

func (self AlertThresholds) StateFor(value AlertThresholdValue) AlertState {
	if value >= self.Critical.Value {
		return &AlertStateCritical{self.Critical.Message}
	} else if value >= self.Warn.Value {
		return &AlertStateWarn{self.Warn.Message}
	} else {
		return &AlertStatePass{}
	}
}

//endregion

type AlertState interface {
	// NotifyMessage - when None, `/resolve` should be called, if Some, `/notify` should be called.
	NotifyMessage() Option[AlertMessage]
}

//region AlertStatePass

type AlertStatePass struct{}

func (self *AlertStatePass) NotifyMessage() Option[AlertMessage] {
	return None[AlertMessage]()
}

//endregion

//region AlertStateWarn

type AlertStateWarn struct {
	message AlertMessage
}

func (self *AlertStateWarn) NotifyMessage() Option[AlertMessage] {
	return Some(self.message)
}

//endregion

//region AlertStateCritical

type AlertStateCritical struct {
	message AlertMessage
}

func (self *AlertStateCritical) NotifyMessage() Option[AlertMessage] {
	return Some(self.message)
}

//endregion
